drivers/disk/flashdisk.c

```diff
@@ -13,14 +14,28 @@
 #include <zephyr/init.h>
 #include <zephyr/device.h>
 #include <zephyr/drivers/flash.h>
+#include <zephyr/storage/flash_map.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(flashdisk, CONFIG_FLASHDISK_LOG_LEVEL);
+
+struct flashdisk_data {
+	struct disk_info info;
+	unsigned int area_id;
+	off_t offset;
+	size_t size;
+	size_t sector_size;
+	size_t page_size;
+};
 
-#define SECTOR_SIZE CONFIG_DISK_FLASH_SECTOR_SIZE
+#define DT_DRV_COMPAT zephyr_flash_disk
 
-static const struct device *flash_dev;
+#define DEFINE_BUFFER(n) uint8_t _buf##n[DT_INST_PROP(n, buffer_size)];
```

Comment:
Making the property optional sounds horrible. It would be better to just get rid of it completely, if the SPI memory case could be handled.

Reply:
Something like
```
#define RMW_BUFFER_SIZE(n) DT_PROP_OR(DT_GPARENT(DT_INST_PROP(n, partition), erase_block_size, SPI_NOR_SECTOR_SIZE)
```

could be used instead but it would be problematic as it assumes that everything that does not provide `erase-block-size` uses hardcoded `SPI_NOR_SECTOR_SIZE` (which can be true right now, but won't necessarily remain true over time).

Reply:
> Something like
> 
> ```
> #define RMW_BUFFER_SIZE(n) DT_PROP_OR(DT_GPARENT(DT_INST_PROP(n, partition), erase_block_size, SPI_NOR_SECTOR_SIZE)
> ```
> 
> could be used instead but it would be problematic as it assumes that everything that does not provide `erase-block-size` uses hardcoded `SPI_NOR_SECTOR_SIZE` (which can be true right now, but won't necessarily remain true over time).

We should not use the `SPI_NOR_SECTOR_SIZE` unless you want to check compatible to figure out whether the device is SPI/QSPI connected.
I think that we should handle two cases here:
 *  `block-size` that handles everything that does not define `erase-block-size` and overrides the `erase-block-size`, as the Kconfig option used to do;
 * fallback to `erase-block-size` so that the default SoC flash info can be used with simplified DTS disk entry.
 
otherwise error, because we do not know what to do.

The `erase-block-size` and `block-size` can be both pulled from DTS.
Getting `SPI_NOR_SECTOR_SIZE` here means leaking definitions from flash drivers directly here.

Another solution, not proposed for this PR, would be to pull the `SPI_NOR_SECTOR_SIZE` out of the header and move it into DTS definition for the SPI connected flash, as `erase-block-size`.

Reply:
We have some naming problems with flash, because the default erase unit is "sector" in SPI NOR, but we call the thing `erase-block-size` in SoC NOR. In SPI NOR there is BLOCK, which is larger than "sector".
Actually erase may start erasing by blocks when given large enough range of sectors.

We hardcode the sizes in spi_nor.h header (both BLOCK and SECTOR), but these values depend of device configuration and can be obtained from SFD either from DTS or at run-time (this does not help us here at all).
So my idea of adding `erase-block-size` to SPI flash devices, no longer sounds that good because, it duplicates what SFD does, and it would have to be array of values.

Reply:
@de-nordic, yes, and on top of that even SoC NOR can have multiple erase-block-sizes, all this complexity is nicely handled (avoided) with the definition of `buffer-size` and letting the user "fine"-tune this values for the specific hardware.